DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp

```diff
@@ -157,44 +120,42 @@ int main() {
         gold[b] = 0;
       }
       for (int i = 0; i < kInitNumInputs; ++i) {
-        int b = input_host[i] % kNumOutputs;
-        gold[b]++;
+        int hist_group = input_host[i] % kNumOutputs;
+        gold[hist_group]++;
       }
     }
 
     // Host accessor is now out-of-scope and is destructed. This is required
     // in order to unblock the kernel's subsequent accessor to the same buffer.
 
-    for (int i = 0; i < kNumRuns; i++) {
-      switch (i) {
-        case 0: {
-          std::cout << "Beginning run without on-chip memory caching.\n\n";
-          Histogram<false>(q, input_buf, output_buf, e);
-          break;
-        }
-        case 1: {
-          std::cout << "Beginning run with on-chip memory caching.\n\n";
-          Histogram<true>(q, input_buf, output_buf, e);
-          break;
-        }
-        default: {
-          Histogram<false>(q, input_buf, output_buf, e);
+    // iterate over the cache depths
+    for (int i = 0; i <= kMaxCacheDepth; i++) {
+
+      std::cout << "Beginning run without cache depth " << i;
+      if (i == 0) { std::cout << " (no cache)"; }
+      std::cout << std::endl;
+
+      fpga_tools::UnrolledLoop<kMaxCacheDepth+1>([&](auto j) {
+        if (j == i) {
+          ComputeHistogram<j>(q, input_buf, output_buf, e);
         }
-      }
+      });
 
-      // Wait for kernels to finish
+      // Wait for kernel to finish
       q.wait();
 
       // Compute kernel execution time
-      t1_kernel = e.get_profiling_info<info::event_profiling::command_start>();
-      t2_kernel = e.get_profiling_info<info::event_profiling::command_end>();
+      t1_kernel = 
+        e.get_profiling_info<sycl::info::event_profiling::command_start>();
+      t2_kernel = 
+        e.get_profiling_info<sycl::info::event_profiling::command_end>();
       time_kernel = (t2_kernel - t1_kernel) / kNs;
 
       // Get accessor to output buffer. Accessing the buffer at this point in
       // the code will block on kernel completion.
-      host_accessor output_host(output_buf, read_only);
+      sycl::host_accessor output_host(output_buf);
```

Comment:
I write to the histogram memory a line or two later to clear it.  I noticed I had an error in my code that the test didn't catch, because the old (still correct) values were still sitting in the output buffer, so I added this to clear the output buffer.